tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php

```diff
@@ -283,6 +286,18 @@ public function testTransactionalReturnValue(): void
         self::assertEquals(42, $res);
     }
 
+    public function testTransactionalDDL(): void
+    {
+        $this->connection->transactional(static function (Connection $conn): void {
+            $table = new Table('foo', [new Column('bar', Type::getType('string'))]);
```

Comment:
This test is incomplete. It performs a single atomic statement that doesn't need to be wrapped in a transaction. Please modify by adding another always failing statement (e.g. `throw new Exception()` after the `foreach`). Then define the expected behavior.

Reply:
I am not testing failing statements, but successful one. MySQL commits operation successfully.

Reply:
The successfull scenario doesn't need a transaction.

Reply:
Successful test scenario needs transaction if it exists to assert no exception has been thrown during successful transaction.

Reply:
DDL statements are not transactional in MySQL and Oracle. No matter how much one wants them to be. If there are two DDLs in a transaction, and the second one fails, the first one will get committed no matter what API is used. I don't really understand this test.

Reply:
It's understandable that if there are two DDLs in a transaction, and the second one fails, the first one will get commited in MySQL. And in that case nobody would blame DBAL that it cannot do a rollback. But again, this test isn't about failure but success. If all operations in `transaction()` block succeed, there is no reason for DBAL to throw exception.

You are saying this doesn't need to be running in transaction for MySQL... But library using DBAL doesn't know what kind of statements user passes in, it will just always use transactions, because there is no point not to. I also need to remind this is not a MySQL test. This test is being executed for all RDBMses. So it does make sense for it to be always running in transaction, because transactional DDL is supported in some of them. Similar rational applies for using this in libraries. 

Now, I was going for minimal test scenario, but if you seriously need hardcoded test case where transaction looks more useful even on MySQL, I could just change the queries in transaction to do some `insert` and _then_ (successful) DDL. In such case when first (and only) DDL fails, changes would be rollbacked even in MySQL, so you would have to stop suggesting using transaction there doesn't make sense.

Reply:
> But again, this test isn't about failure but success.

This is fundamentally wrong. There's no point in running a transaction if you don't care about error handling. The whole point of a transaction is error handling (well, and isolation).

>  If all operations in `transaction()` block succeed, there is no reason for DBAL to throw exception.

The method assumes that it controls the transaction flow but it fails to. Hence the exception.